common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -410,7 +412,9 @@ public boolean endsWith(final UTF8String suffix, int collationId) {
     if (collationId == CollationFactory.LOWERCASE_COLLATION_ID) {
       return this.toLowerCase().endsWith(suffix.toLowerCase());
     }
-    return matchAt(suffix, numBytes - suffix.numBytes, collationId);
+    if (suffix.numBytes == 0 || this.numBytes == 0) return suffix.numBytes==0;
+
+    return CollationFactory.getStringSearch(this, suffix, collationId).last()==this.numChars()-suffix.numChars();
```

Comment:
for clarity, I think we can separate this out into `private boolean collatedEndsWith`